File: bindings/swift/examples/MuxRoundtrip.swift
// MuxRoundtrip.swift
//
// ASPIRATIONAL EXAMPLE: no `Mediaway` Swift package exists yet — no
// `mediaway-container-ffi` C ABI crate has shipped either. This file shows
// the target ergonomics for a future Swift binding over Mediaway's C ABI: a
// C bridging header wrapped in idiomatic Swift (classes with `deinit`/
// explicit `close()`, `throws`/`try` instead of raw C status codes). See
// ../README.md and docs/spec/c-ffi.md.
//
// Mirrors examples/mux_roundtrip.rs: register one H.264 video track and one
// AAC audio track, push fake packets for a simulated 3-second clip, flush,
// and read the fragmented MP4 bytes back with a streaming demuxer.
//
// Run (once the real package exists):
//     swift run MuxRoundtrip

import Mediaway
import Foundation

private let frameCount = 90 // 3 s at 30 fps
private let keyframeInterval = 30
private let videoTimeBase = Rational(1, 30)
private let audioTimeBase = Rational(1, 48_000)

/// Mux one video + one audio track into fragmented MP4 bytes.
func buildFmp4() throws -> Data {
    let muxer = Muxer()

    // 1. Register tracks (open state).
    let videoId = try muxer.addTrack(.video(
        codec: .h264,
        timeBase: videoTimeBase,
        width: 1920,
        height: 1080,
        extraData: Data()
    ))
    let audioId = try muxer.addTrack(.audio(
        codec: .aac,
        timeBase: audioTimeBase,
        extraData: Data(),
        sampleRate: 48_000,
        channels: 2
    ))

    // 2. Transition to a live session — track registration closes here.
    let session = try muxer.begin()
    defer { session.close() }

    for i in 0..<frameCount {
        try session.pushPacket(Packet(
            streamId: videoId,
            pts: Int64(i),
            dts: Int64(i),
            duration: 1,
            isKeyframe: i % keyframeInterval == 0,
            isDiscard: false,
            payload: Data([0x00, 0x00, 0x00, 0x01]) // placeholder NAL unit
        ))
        try session.pushPacket(Packet(
            streamId: audioId,
            pts: Int64(i) * 1_600,
            dts: Int64(i) * 1_600,
            duration: 1_600,
            isKeyframe: true,
            isDiscard: false,
            payload: Data([0xFF, 0xF1])
        ))
    }
    try session.flush()

    // 3. Pull bytes — caller owns I/O, the muxer never touches disk.
    return session.pollBytes()
}

/// Feed muxed bytes into a demuxer and count video vs. audio packets.
func demuxAndCount(_ data: Data) throws -> (video: Int, audio: Int) {
    let demuxer = Demuxer()
    defer { demuxer.close() }

    try demuxer.pushBytes(data)

    let streams = demuxer.streams
    print("mux_roundtrip: demuxer sees \(streams.count) stream(s)")
    for stream in streams {
        if let geometry = stream.geometry {
            print("  stream \(stream.id) — \(stream.codec) \(geometry.width)x\(geometry.height)")
        } else {
            print("  stream \(stream.id) — \(stream.codec) (no geometry)")
        }
    }

    var videoCount = 0
    var audioCount = 0
    while let packet = try demuxer.pollPacket() {
        guard let stream = streams.first(where: { $0.id == packet.streamId }) else { continue }
        if stream.codec == .h264 {
            videoCount += 1
        } else {
            audioCount += 1
        }
    }
    return (videoCount, audioCount)
}

// MARK: - Entry point

do {
    let fmp4Bytes = try buildFmp4()
    print("mux_roundtrip: \(frameCount) frames -> \(fmp4Bytes.count) bytes of fMP4")

    let (videoCount, audioCount) = try demuxAndCount(fmp4Bytes)
    print("mux_roundtrip: recovered \(videoCount) video + \(audioCount) audio packets")
    assert(videoCount > 0)
    print("mux_roundtrip: OK")
} catch {
    print("mux_roundtrip: failed — \(error)")
    exit(EXIT_FAILURE)
}

File: bindings/kotlin/examples/MuxRoundtrip.kt
/*
 * Mux + demux roundtrip — aspirational quick-start example.
 *
 * ASPIRATIONAL EXAMPLE: no `mediaway` Kotlin/Java package exists yet. This file
 * shows the target ergonomics for a future Kotlin binding over Mediaway's C ABI
 * (JNI under the hood, wrapped idiomatically: Closeable resources used with
 * `.use { }`, Kotlin exceptions instead of raw error codes). The API is also
 * meant to read naturally from plain Java, so no Kotlin-only surface (inline
 * value classes, etc.) is used in the public shape. See ../README.md and
 * docs/spec/c-ffi.md.
 *
 * Mirrors examples/mux_roundtrip.rs: register one H.264 video track and one
 * AAC audio track, push fake packets for a simulated 3-second clip, flush,
 * and read the fragmented MP4 bytes back with a streaming demuxer.
 *
 * Run (once the real package exists):
 *     kotlinc MuxRoundtrip.kt -include-runtime -d mux-roundtrip.jar
 *     java -jar mux-roundtrip.jar
 */

import io.mediaway.container.AudioStreamInfo
import io.mediaway.container.Codec
import io.mediaway.container.Demuxer
import io.mediaway.container.Muxer
import io.mediaway.container.Packet
import io.mediaway.container.Rational
import io.mediaway.container.VideoStreamInfo

private const val FRAME_COUNT = 90 // 3 s at 30 fps
private const val KEYFRAME_INTERVAL = 30
private val VIDEO_TIME_BASE = Rational(1, 30)
private val AUDIO_TIME_BASE = Rational(1, 48_000)

/** Mux one video + one audio track into fragmented MP4 bytes. */
fun buildFmp4(): ByteArray {
    val muxer = Muxer()

    // -- 1. Register tracks (open state) --------------------------------
    val videoId = muxer.addTrack(
        VideoStreamInfo(
            codec = Codec.H264,
            timeBase = VIDEO_TIME_BASE,
            width = 1920,
            height = 1080,
            extraData = ByteArray(0),
        )
    )
    val audioId = muxer.addTrack(
        AudioStreamInfo(
            codec = Codec.AAC,
            timeBase = AUDIO_TIME_BASE,
            extraData = ByteArray(0),
            sampleRate = 48_000,
            channels = 2,
        )
    )

    // -- 2. Transition to a live session — track registration closes here --
    muxer.begin().use { session ->
        for (i in 0 until FRAME_COUNT) {
            session.pushPacket(
                Packet(
                    streamId = videoId,
                    pts = i.toLong(),
                    dts = i.toLong(),
                    duration = 1,
                    isKeyframe = i % KEYFRAME_INTERVAL == 0,
                    isDiscard = false,
                    payload = byteArrayOf(0x00, 0x00, 0x00, 0x01), // placeholder NAL unit
                )
            )
            session.pushPacket(
                Packet(
                    streamId = audioId,
                    pts = i.toLong() * 1_600,
                    dts = i.toLong() * 1_600,
                    duration = 1_600,
                    isKeyframe = true,
                    isDiscard = false,
                    payload = byteArrayOf(0xFF.toByte(), 0xF1.toByte()),
                )
            )
        }
        session.flush()

        // -- 3. Pull bytes — caller owns I/O, the muxer never touches disk --
        return session.pollBytes()
    }
}

/** Feed muxed bytes into a demuxer and count video vs. audio packets. */
fun demuxAndCount(data: ByteArray): Pair<Int, Int> {
    Demuxer().use { demuxer ->
        demuxer.pushBytes(data)

        val streams = demuxer.streams
        println("mux_roundtrip: demuxer sees ${streams.size} stream(s)")
        for (stream in streams) {
            val geometry = stream.geometry
            if (geometry != null) {
                println("  stream ${stream.id} — ${stream.codec} ${geometry.width}x${geometry.height}")
            } else {
                println("  stream ${stream.id} — ${stream.codec} (no geometry)")
            }
        }

        var nVideo = 0
        var nAudio = 0
        while (true) {
            val packet = demuxer.pollPacket() ?: break
            val stream = streams.first { it.id == packet.streamId }
            if (stream.codec == Codec.H264) nVideo++ else nAudio++
        }
        return nVideo to nAudio
    }
}

fun main() {
    val fmp4Bytes = buildFmp4()
    println("mux_roundtrip: $FRAME_COUNT frames -> ${fmp4Bytes.size} bytes of fMP4")

    val (nVideo, nAudio) = demuxAndCount(fmp4Bytes)
    println("mux_roundtrip: recovered $nVideo video + $nAudio audio packets")
    check(nVideo > 0)
    println("mux_roundtrip: OK")
}
